core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java

```diff
@@ -274,12 +277,24 @@ public RexNode visitLocalRef(RexLocalRef localRef) {
       final RexInputRef ref = (RexInputRef) refToWindow.get(index);
       projectList.add(ref);
     }
-
-    return relBuilder.push(window)
+    RelNode sorted = requireSorted(window,
```

Comment:
Why is this added?

Reply:
In the first commit of the PR the Sort has been added in WindowedAggRelSplitter, the solution is not good because it might produce Sort-Project-Window, which causes failure of PlannerTest#testDuplicateSortPlanWithOver.
The correct result should be Project-Sort-Window, so I move the change from WindowedAggRelSplitter to here.

Reply:
It still does not answer the question why do you need to add a sort node here.

Reply:
If you go with "LogicalWindow has empty collation" approach, then you should not add Sort nodes. What's the purpose? Collation is empty after all.

Reply:
For example, If we convert
```
LogicalProject[collation: [0]]
|-LogicalProject[collation: [0]]
```
to
```
LogicalProject[collation: [0]]
|-LogicalWindow[collation: []]
```
with no Sort node added, If the implementation of LogicalWindow has empty collation too, then the node `LogicalProject[collation: [0]]` can not be implemented because there is no input which is with a suitable collation.

Reply:
> But the converted Rel's collation should always match the original Rel's

This is false.

Rel can alter collation as it wishes.
For instance, Sort rel has different collation than its input.

Reply:
I suggest to set EMPTY collation for Project nodes that have OVER expressions.

Reply:
See `c)` in TLDR section of the comment:  https://issues.apache.org/jira/browse/CALCITE-2648?focusedCommentId=16734082&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16734082

Reply:
> But the converted Rel's collation should always match the original Rel's

I understand your point. 

What if adding a rule to add a Sort node on top of the LogicalWindow?

Reply:
It must not be done

Reply:
> See c) in TLDR section of the comment
I understand your suggestion.

And could you please explain a bit why it is better to make window function call have no order assurance? Is that a trade-off?
If window function call could not preserve order, why a simple project should preserve order? Is there some principles of this?

Reply:
> Can you please clarify why don't just just update collation for LogicalProject?
I think I should have a reason to change the current ordering behavior of Calcite. There might be some downstream projects implementing there ordered window call.

Reply:
Technically speaking, I think `LogicalWindow(window=over(partition by deptno order by empid))` should have a collation of `collation=[0:CLUSTERED, 1:ASC]`. I believe, most of the implementation would sort the rows as it is prescribed by `order by`, so it makes sense to have those properties even at logical level, so the optimizer can remove redundant sorts, etc. I also think that CalcRelSplitter should create multiple LogicalWindow rels in case windows are different (e.g. if they have incompatible `partition by` clauses).

However @julianhyde believes `LogicalWindow` should have empty collation (see his comments in CALCITE-2648). I'm sure it loses lots optimization possibilities (e.g. FilterWindowTranspose), however I don't care to spend time on discussing collation for LogicalWindow as very few people care to discuss CALCITE-2648 / CALCITE-2593.

Reply:
>And could you please explain a bit why it is better to make window function call have no order assurance? 

In case the "agreed" change is to make "LogicalWindow" to always have "empty collation", then LogicalProject with OVER expressions should better have EMPTY collation as well in order to keep consistency between different relations. In other words, if we suppose that "OVER resets collation to an unknown state" (which what means empty collation for LogicalWindow), then we should specify the same for "LogicalProject with OVER expressions" (==collation of LogicalProject that have OVER should be empty)

Reply:
> There might be some downstream projects implementing their ordered window call.

 Implementing `LogicalProjects` that have OVER expressions is a non-trivial task. It almost sure would require to have logic like CalcRelSplitter since single LogicalProject might happen to have OVER(partition by=row_number() OVER(...) ) kind of nesting, so it is not trivial to implement.

Note that regular databases (e.g. PostgreSQL) do not care to keep collations when computing OVER expressions. I don't care to read SQL standard, however I'm almost sure standard says that OVER might alter output order of rows, so my bet is standard says LogicalProject(x=count() over (order by y), input=sort(table_scan, order by=z)) might produce whatever sort order.

Reply:
Sorry for the late reply, I've been working on some other issues these days.

>   I believe, most of the implementation would sort the rows as it is prescribed by order by, so it makes sense to have those properties even at logical level, so the optimizer can remove redundant sorts, etc.

@vlsi - Wouldn't having the properties in Enumerable level or user-defined convention level helps optimizer remove redundant sorts as well? SortRemoveRule doesn't force its input to be a logical node.

> In case the "agreed" change is to make "LogicalWindow" to always have "empty collation", then LogicalProject with OVER expressions should better have EMPTY collation as well in order to keep consistency between different relations.

Agreed. And I would write separate methods for logical window and enumerable window in RelMdCollation, returning empty collation/implementation collation respectively. Of course we should achieve agreement on whether to wipe LogicalWindow's collation before doing that.